New module: openstack_mng

morsik opened the pull request:
### What does this PR do?

Adds new module.
### Tests written?

No
### Module info

This module allows management of OpenStack services (using `openstack-service` command).

Additionally, it allows to restart service only when it's running longer than specified time (depends on systemd; if systemd is not used on system, it'll be restarted immediately). This can be useful when you use this with Salt Reactor and more than 1 script/daemon/whatever would try to restart the same service in case of some self-healing.
#### Why such module?

`service.restart` allows to restart service by name (like `service.restart openstack-nova-compute`). But `openstack-service` command can restart all OpenStack services that works on specific host.

For example:
`openstack-service restart nova` will restart `nova-compute` if it's Compute Node, or `nova-conductor`, `nova-scheduler`, `nova-conductor` and others if it's Controller Node.

Name for module is chosen on existing pattern (there is already `openstack_config`), and it can be used later for other management stuff.

Commit:
New module: openstack_mng

cachedout commented on salt/modules/openstack_mng.py:
```diff
@@ -0,0 +1,114 @@
+
+# Import python libs
+import logging
+import time
```
This appears to be unused. If so, could you please remove it?

morsik replied:
Removed. I was using it previously, but changed and forgot about it ;)

morsik committed:
Removed unused module time

morsik commented:
Removed. I was using it previously, but changed and forgot about it ;)

cachedout commented on salt/modules/openstack_mng.py:
```diff
@@ -0,0 +1,113 @@
+    '''
+    Only load this module if openstack-service is installed
+    '''
+    if HAS_OPENSTACK:
```
We can just replace the block above with `os.path.isfile('/usr/bin/openstack-service')` here and reduce the code a bit.

morsik replied:
Fixed.

cachedout commented on salt/modules/openstack_mng.py:
```diff
@@ -0,0 +1,113 @@
+        salt '*' openstack_mng.stop_service neutron
+    '''
+
+    ret = __salt__['cmd.run_all'](['/usr/bin/openstack-service', 'stop', service_name])
```
`cmd.retcode` instead?

morsik replied:
Fixed.

cachedout commented on salt/modules/openstack_mng.py:
```diff
@@ -0,0 +1,113 @@
+        for service in services:
+            service_info = __salt__['service.show'](service)
+
+            boot_time = float(open('/proc/uptime').read().split(' ')[0])
```
Please use `salt.utils.fopen` here instead of `open`.

morsik replied:
Fixed.

But after reading about `salt.utils.fopen` and `CLOEXEC` flag, could you exaplain situation when fd would leak? Python automatically closes fd when using inline `.read()` expression.

cachedout commented on salt/modules/openstack_mng.py:
```diff
@@ -0,0 +1,113 @@
+            boot_time = float(open('/proc/uptime').read().split(' ')[0])
+
+            expr_time = int(service_info.get('ExecMainStartTimestampMonotonic', 0)) / 1000000 < boot_time - minimum_running_time
+            expr_active = True if service_info.get('ActiveState') == "active" else False
```
This can  be `expr_active = service_info.get('ActiveState') == "active"`

morsik replied:
Ahh, right! Fixed.

morsik committed:
Removed redundant openstack-service binary condition

morsik committed:
Removed redundant expr_active condition result

morsik committed:
Use salt wrapper for open

morsik committed:
Use simpler cmd.retcode instead of cmd.run_all

Since `cmd.retcode` can return `int` or `None` I must explicity check if it's `0`. `if not` checks for `0` *and* `None`.

morsik commented:
It is better now? ;)

cachedout commented on salt/modules/openstack_mng.py:
```diff
@@ -0,0 +1,107 @@
+        salt '*' openstack_mng.start_service neutron
+    '''
+
+    if __salt__['cmd.retcode'](['/usr/bin/openstack-service', 'start', service_name]) == 0:
```
It would be cleaner and shorter if we changed this logic in the functions that you've written to something more like this:

`return __salt__['cmd.retcode'](['/usr/bin/openstack-service', 'start', service_name]) == 0`

morsik replied:
Actually I should split that into more lines cause it exceeds 80 columns. Or we don't care about that? If we don't care I'll change that.

Or if we care, I can change that to:

```python
    os_cmd = ['/usr/bin/openstack-service', 'start', service_name]
    return __salt__['cmd.retcode'](os_cmd) == 0
```

cachedout replied:
Our 80-char limit is a preference but not enforced. Either way is fine with me. :]

cachedout commented:
Hi @morsik I just left one comment in-line. I think we can make this just a little cleaner. Have a look and let me know what you think. Thanks!

morsik committed:
Use shorter return with condition

morsik commented:
Check out now ;)

cachedout commented on salt/modules/openstack_mng.py:
```diff
@@ -0,0 +1,101 @@
+        return (False, 'The openstack-service binary could not be found.')
+
+
+def start_service(service_name):
```
I wonder if we should simplify these function names down to just start, stop and so on. Thoughts?

morsik replied:
We could, but I left them that, cause this module could be used for other OpenStack management stuff, and it could lead to `start_something_else` later…

cachedout replied:
Fair enough. Just wanted to mention it. Thanks!
